### PR TITLE
grain: update assembly and make accumulate inlinable

### PR DESCRIPTION
**Pull request by ericlagergren**
Signed-off-by: Eric Lagergren <eric@ericlagergren.com>

### CODE REVIEW

**Comment by ericlagergren** [comment]
cc: @greatroar

**Comment by ericlagergren** [comment]
On my x86_64 MBP I get ~124 MB/s using assembly, ~95 MB/s using Go.

On my M1 MacBook Air I get ~150 MB/s using Go (and 102 MB/s using assembly through Rosetta).

File: grain/asm/asm.go [hunk 1]
```diff
@@ -263,56 +262,40 @@ func declareAccumulate() {
...
-		XORW(rem.As16(), acctmp.As16())
+		Comment("acctmp ^= uint16(regtmp) & uint16(mask)")
+		ANDW(regtmp.As16(), mask.As16()) // clobber mask
+		XORW(mask.As16(), acctmp.As16())
 
 		Comment("reg >>= 1")
 		shrq(1, reg)
 
+		Comment("pt >>= 1")
+		shrl(1, pt)
```
Comment (greatroar) [inline grain/asm/asm.go:286]: This can go in the if block, below.

**Comment by greatroar** [comment]
```
Auth-8               109MB/s ± 1%
AuthGeneric-8       96.7MB/s ± 1%
```

The question is whether it's the assembly doing the magic now, though, or just the unrolling of the loop.

**Comment by ericlagergren** [comment]
It's very likely loop unrolling. The instructions are nearly the same. The only additional improvement I can immediately think of is reordering the instructions to eliminate false dependencies or stalls or whatever.

Unrolling the Go code should make it equally quick, I think. Could generate it with a small program.

**Pull request merged**